TST: openblas for Azure MacOS

tylerjereddy opened the pull request:
**Objective**: test the MacOS configuration that most closely matches what many users have access to (wheels built with `openblas`) & perhaps slightly reduce the testing dissonance between development CI & numpy-wheel CI builds, by building with `openblas`.

<details>

[Reference to `brew install openblas`](https://github.com/matthew-brett/multibuild/blob/devel/library_builders.sh#L99) in `multibuild` repo

Previously, in [a master branch run](https://dev.azure.com/numpy/numpy/_build/results?buildId=1024&view=logs) for MacOS / Python 3.6 Azure: `7218 passed, 39 skipped, 10 xfailed, 1 xpassed in 309.53 seconds`

Azure CI result for MacOS with openblas in this branch should be: `7217 passed, 40 skipped, 10 xfailed, 1 xpassed in 298.79 seconds`
  - 1 extra test skip related to `xerbla` & not clear if this is skipped in our matching Python 3.6 MacOS [wheel build log](https://api.travis-ci.org/v3/job/460421265/log.txt) where test skips aren't verbose, but far more tests are skipped (`167` to `170`).

The `Build NumPy` stage should now clearly show `openblas` as `FOUND` by distutils, and the end of this stage should no longer print the warnings about various missing linalg acceleration libraries.


</details>

rgommers commented on azure-pipelines.yml:
```diff
@@ -54,19 +54,25 @@ jobs:
     displayName: 'make gfortran available on mac os vm'
+  # we currently seem to use homebrew for openblas in wheels
+  # infrastructure; at the time of writing this means
+  # openblas 0.3.3 on OSX is used vs. Linux clearly targeted at 0.3.0
```
Can we clarify if this is for user install via Homebrew, in `numpy-wheels` for our releases, or both? IIRC it's Homebrew only, and our wheels have 0.3.0?

tylerjereddy replied:
I think I'm confused because, as noted in details drop-down above, I see [`brew install openblas`](https://github.com/matthew-brett/multibuild/blob/devel/library_builders.sh#L99) in the `multibuild` (submodule) repo, but I guess that is maybe not the repo I should be looking at.

tylerjereddy replied:
the `on_openblas.rst` in numpy-wheels suggests I may need to check another repo instead: https://github.com/MacPython/openblas-libs

rgommers replied:
It's a bit unclear indeed.

- For Windows there's hardcoded SHA's of the OpenBLAS used that can be checked (in `appveyor.yml`).
- Both the NumPy and SciPy builds use an environment variable for Linux and macOS it looks like: https://github.com/MacPython/numpy-wheels/blob/master/env_vars.sh#L2, https://github.com/MacPython/scipy-wheels/blob/master/env_vars.sh#L2 (bothset to `v0.3.0` currently)

rgommers replied:
Here is an example of changing the version that is used: https://github.com/MacPython/numpy-wheels/commit/3d0da6b6acac050e67509ed6a469dba1602b1890

tylerjereddy replied:
Ralf was right here -- I now correctly pull in the MacPython pre-built OpenBLAS binary in the workflow. Marking as resolved.

rgommers commented on azure-pipelines.yml:
```diff
@@ -54,19 +54,25 @@ jobs:
-  # LAPACK configuration on Azure macos image; at the time of writing
-  # this plagues homebrew / macports NumPy builds, but we will
-  # circumvent for now by aggressively disabling acceleration for
-  # macos NumPy builds / tests; ACCELERATE=None on its own is not
```
We haven't discussed dropping Accelerate support, but I'd be fine doing that now or later on (following SciPy). Especially if we don't use Accelerate in CI.

tylerjereddy replied:
I'll mark this as resolved -- I can't remember the status of Accelerate deprecation, but it isn't directly connected to using OpenBLAS vs. no acceleration at all here, which is obviously an improvement.

tylerjereddy commented:
Adding WIP tag until the openblas version thing is known for sure.

tylerjereddy commented:
I'm still standing by my original analysis that we use `homebrew` to install `openblas` for Mac wheel builds, and it ignores any environment variables we set to control version of `openblas` used, so my comment above still holds as far as I can tell.

I'll put my detailed control flow analysis below the fold in details section.

<details>

- `.travis.yml` in numpy-wheels repo has `before_install` section that presumably sets things up before the wheel build happens in install
- the `before_install` command in the `before_install` section of numpy-wheels repo `.travis.yml` is actually defined in `travis_osx_steps.sh` in multibuild repo, which is selectively sourced on OSX platform by `source multibuild/travis_steps.sh` in `.travis.yml` in numpy-wheels repo
- `multibuild/travis_osx_steps.sh` also sources `multibuild/library_builders.sh`
- in `multibuild/library_builders.sh`: 
    - an OpenBLAS base URL is set early on with variable `GF_LIB_URL`
    - then `OPENBLAS_VERSION="${OPENBLAS_VERSION:-0.2.18}" `
      - so, basically feed through the `OPENBLAS_VERSION` variable, if defined, which it is at `0.3.0`
    - the farther down there is the function `build_openblas()`
        - it clearly uses `brew install openblas` & **ignores** `OPENBLAS_VERSION` for OSX, as I've already indicated
- bash `source` command in `multibuild/travis_osx_steps.sh`  will actually **execute** all the functions in `multibuild/library_builders.sh`
  - this includes `build_openblas`, which does `brew install openblas` and clearly ignores the `OPENBLAS_VERSION` environment variable used for Linux openblas
</details>

rgommers commented:
> @mattip suggests I may be able to confirm the openblas version somehow in the downloaded wheels -- I should check for a recent cron job wheel build if possible to confirm

I tried to do that yesterday, but couldn't find an obvious way (but maybe you'll have more luck). The `__config__.py` file only confirms that the build picked up `openblas`, but no version. I also couldn't find a function in OpenBLAS itself to query the version.

Actually extending `system_info.py` so version numbers do land in `__config__.py` would be super useful.

charris commented:
There are `openblas` libs for Darwin at http://wheels.scipy.org/

tylerjereddy commented:
> Actually extending system_info.py so version numbers do land in __config__.py would be super useful.

Yeah tagging the wheel somehow with openblas version just came up in the [weekly meeting notes](https://hackmd.io/Gn1ymjwkRjm9WVY5Cgbwsw?both)

tylerjereddy commented:
I'm saying we use `brew install openblas` which means this recipe: https://github.com/Homebrew/homebrew-core/blob/master/Formula/openblas.rb

which uses `url "https://github.com/xianyi/OpenBLAS/archive/v0.3.3.tar.gz"`

tylerjereddy commented:
On the other hand -- `Error: No available formula with the name "openblas" ` when I injected `brew info openblas` in the [travis build](https://travis-ci.org/MacPython/numpy-wheels/jobs/461063029) for Python 3.6 -- some other osx builds error for another reason with that same change.

tylerjereddy commented:
Ah, I think I see the flaw in my control flow above -- the last line in `../multibuild/travis_osx_steps.sh` sources a config file (`config.sh `) that overrides `build_libs` to treat osx and linux the same, so that's probably why we're getting `0.3.0` from my dissection above.

tylerjereddy commented:
I'll restore the WIP tag and try to replicate what the wheel builds do here then -- that's certainly pretty messy!

tylerjereddy commented:
This is still confusing me since `build_openblas` would still be called with brew -- there must be another interception, maybe.

I wonder about the design decision to use multiple cross-sourcing bash scripts that sometimes have functions with the same names.

charris commented:
I think the whole thing just growed :) It's a good example of the precept that, in practice, usefulness trumps beauty. Could certainly use a cleanup at some point, maybe if/when we decide to move building to azure.

charris commented:
@matthew-brett Can you help?

tylerjereddy commented:
Updated to reflect the best match to our pre-built openblas wheels based on an [analysis script](https://github.com/tylerjereddy/hash_search) that scrapes the wheels URL for candidates. Also, rolled back gfortran install to 4.9 to better match wheel toolchain.

However, the WIP tag must remain despite passing build / tests as the f2py tests are not capable of detecting a C compiler despite clear presence of clang (aliased to gcc as well) for building after making these adjustments (the brew option was much easier!).

Manually linking the brew-installed gcc-4.9 to /usr/local/lib is probably hairy. The wheels toolchain uses some old Mac `.dmg` file instead.

Future versions of openblas will have version tagging to make the detective work a little easier, though I'm not really clear on if it will truly help us if we move our openblas builds to specific commit hashes outside of major releases, as was the case here.

Certainly it seems that we can narrow down the possible pre-built openblas binaries used with a bit of effort in the wheel "post-mortem." Could also try debug printing the URLs used in the wheel toolchain, but that control flow is hard to follow.

tylerjereddy committed:
TST: openblas for Azure MacOS

* use pre-built openblas binary
for MacOS Azure CI builds to more
closely match the build scenario
used by numpy-wheels

* similarly, roll back gfortran
version to 4.9 to better match
wheel builds

tylerjereddy commented:
This should now be working as intended based on the updates I've pushed from my fork. The final key was to symlink `gfortran-4.9` to `gfortran`--f2py can't handle the case where only the former is present.

In short, Azure Mac OS CI should now use the OpenBLAS backend instead of NO linear algebra backend at all--forming a much closer match to our Mac OS wheels & providing us with a platform to have OpenBLAS version checked for all 3 platforms in Azure CI.

For the reviewer(s): maybe good to quickly confirm that the test result is almost identical to master branch for Mac--just 1 additional skip now, I think. Also, a quick scan to check that OpenBLAS is detected by distutils in the build log now & not on master.

rgommers commented:
Sounds good. I'll try to review tomorrow.

rgommers commented:
This all looks like it works, copying the relevant parts of the CI logs here:

```
2019-03-14 01:59:02 (74.7 MB/s) - ‘openblas-v0.3.3-186-g701ea883-macosx_10_9_intel-gf_1becaaa.tar.gz’ saved [21938775/21938775]

x usr/local/include/cblas.h
x usr/local/include/f77blas.h
x usr/local/include/openblas_config.h
x usr/local/include/lapacke.h
x usr/local/include/lapacke_config.h
x usr/local/include/lapacke_mangling.h
x usr/local/include/lapacke_utils.h
x usr/local/lib/libopenblas.a
x usr/local/lib/libopenblas.dylib
x usr/local/lib/libopenblasp-r0.3.5.dev.a
x usr/local/lib/libopenblasp-r0.3.5.dev.dylib
x usr/local/lib/cmake/openblas/
x usr/local/lib/cmake/openblas/OpenBLASConfig.cmake
x usr/local/lib/cmake/openblas/OpenBLASConfigVersion.cmake
cp: ./usr/local/lib/cmake is a directory (not copied).
##[section]Finishing: install pre-built openblas
```

```
  FOUND:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]

  FOUND:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]

openblas_lapack_info:
customize UnixCCompiler
customize UnixCCompiler
customize UnixCCompiler
C compiler: /usr/bin/clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/usr/local/opt/openssl/include
```

rgommers commented:
Merging, thanks @tylerjereddy
